src/session.c

```diff
@@ -735,6 +747,17 @@ static CK_RV token_login(P11PROV_SESSION *session, P11PROV_URI *uri,
                         pinlen);
     if (ret == CKR_USER_ALREADY_LOGGED_IN) {
         ret = CKR_OK;
+    } else {
+        if (pin && ret == CKR_PIN_INCORRECT) {
+            /* mark this pin as bad or we may end up locking the token */
+            OPENSSL_free(slot->bad_pin);
```

Comment:
I think the pin should be considered as a sensitive data (regardless if it is good or bad). So it should be probably wiped from the memory before freeing here as well as on cleanup (do not see it freed on the session cleanup either).
One more consideration might be to store sha digest of the pin only here to avoid keeping the pin in memory for long time if not necessary.

Reply:
This is done only for pins already stored in memory, so I think hashing is a bit excessive?